File: 2do Curso/1T Programacion Avanzada/Practicas/Programacio Avançada/Examen1/src/Examen1/ContenidorBrossa.java
package Examen1;

public abstract class ContenidorBrossa implements Comparable<ContenidorBrossa> {
	private int color; // color del contenidor
	private Ubicacio ubicacio; // Si està al magatzté un valor de null
	private String codi; // identificador
	private int any; // emmagatzema l'any de compra
	protected float tara;
	public final static int groc = 0;
	public final static int marro = 1;
	public final static int gris = 2;
	public final static int verd = 3;
	public final static int blau = 4;
	public final static int nou = 0;
	public final static int seminou = 1;
	public final static int vell = 2;

	public ContenidorBrossa(int color, String codi, float tara, Ubicacio on, int any) {
// Suposa que el valor de tots els paràmetres és correcte.
// Si on és null s’emmagatzema al magatzem de la població
		this.codi = codi;
		this.color = color;
		this.ubicacio = on;
		this.tara = tara;
		this.any = any;
	}

	public int getEstat() {
		if (any < 2000)
			return vell;
		if (any < 2015)
			return seminou;
		return nou;
	}

	public String getCodi() {
		return codi;
	}

	public String getTipusBrossa() {
		switch (color) {
		case marro:
			return "organic";
		case gris:
			return "rebuig";
		case blau:
			return "paper";
		case groc:
			return "plastic";
		case verd:
			return "vidre";
		default:
			return "ERROR";
		}
	}

	public Ubicacio getUbicacio() {
		return this.ubicacio;
	}

	public String toString() {
		String s = "Codi: " + codi;
		if (ubicacio != null)
			s += "Ubicació: " + ubicacio.toString();
		else
			s += "Ubicació: retirat";
		s += "Tipus brossa " + getTipusBrossa();
		s += "Tara: " + tara;
		return s;
	}

	public int compareTo(ContenidorBrossa o) {
		return codi.compareTo(o.codi);
	}

	public abstract void buidat(double quilos);
}

File: 2do Curso/1T Programacion Avanzada/Practicas/Programacio Avançada/Practica2MiquelRodriguezSergioBenages/src/ContenidorBrossa.cpp
#include "ContenidorBrossa.h"
#include <time.h>
#include <string.h>

ContenidorBrossa::ContenidorBrossa(string codi, int color, string ubicacio, int anyColocacio, float tara)
{
    this->codi = codi;
    if (color != ContenidorBrossa.GROC || color != ContenidorBrossa.MARRO || color != ContenidorBrossa.VERD || color != ContenidorBrossa.GRIS || color != ContenidorBrossa.BLAU)
    {
        throw new exception("Color erroni");
    } else this->color = color;
    this->ubicacio = ubicacio;
    if(anyColocacio < 1000 || anyColocacio > 2022) {
        throw new exception("Any erroni");
    } else this->anyColocacio = anyColocacio;
    this->anyRetirada = 0;
    this->tara = tara;
}

ContenidorBrossa::ContenidorBrossa(string codi, int color, int anyColocacio, float tara)
{
    this(codi, color, NULL, anyColocacio, tara);
}

string ContenidorBrossa::getTipusBrossa()
{
    switch(color)
    {
        case ContenidorBrossa.GROC:
            return "Contenidor plàstic";
            break;
        case ContenidorBrossa.MARRO:
            return "Contenidor orgànic";
            break;
        case ContenidorBrossa.VERD:
            return "Contenidor vidre";
            break;
        case ContenidorBrossa.GRIS:
            return "Contenidor rebuig";
            break;
        case ContenidorBrossa.BLAU:
            return "Contenidor paper";
            break;
    }
}

void ContenidorBrossa::retirarViaPublica()
{
    time_t now;
    struct tm *now_tm;
    now = time(NULL);
    now_tm = localtime(&now);
    int year = now_tm->tm_year + 1900;

    this->anyRetirada = year;
    if (this->ubicacio == NULL)
    {
        throw new exception("El contenidor no està ubicat a la via pública");
    } else this->ubicacio = NULL;
    this->anyColocacio = 0;
}

string ContenidorBrossa::getUbicacio()
{
    if (this->ubicacio == NULL)
    {
        throw new exception("El contenidor està al magatzem");S
    } else return this->ubicacio;
}

void ContenidorBrossa::setUbicacio(string ubicacio)
{
    if (this->ubicacio == NULL)
    {
        retirarViaPublica();
    } else this->ubicacio = ubicacio;
}

int ContenidorBrossa::getCodi()
{
    return this->codi;
}

string ContenidorBrossa::getEstat()
{
    time_t now;
    struct tm *now_tm;
    now = time(NULL);
    now_tm = localtime(&now);
    int year = now_tm->tm_year + 1900;
    int any = year - this->anyColocacio

    if (any > 5)
    {
        return "C   ontenidor vell";
    } else if (any <= 5 && any >= 3)
    {
        return "Contenidor seminou";
    } else if (any < 3)
    {
        return "Contenidor nou";
    } else if (this->anyColocacio == 0)
    {
        return "Contenidor retirat";
    }
}

bool ContenidorBrossa::operator==(ContenidorBrossa *p)
{
    if(this->getCodi() == *p->getCodi()) return true;
    else return false;
}

bool ContenidorBrossa::operator<(ContenidorBrossa *p)
{
    if(this->getCodi < *p->getCodi()) return true;
    else return false;
}

bool ContenidorBrossa::operator>(ContenidorBrossa *p)
{
    return !this->operator<(*p);
}

void toString()
{
    String resultat = "";
    resultat = "Codi: " + this->getCodi() + endl + "Color: " + this->getTipusBrossa() + endl + "Ubicació: ";
    if (this->getUbicacio() == 0)
    {
        resultat += ;
    } else resultat += this->getUbicacio();
    resultat += endl + "Tara: " + this->tara;
}

/*
void buidat (float pes)
{
    // NO ENTENC AQUEST METODE
    this->tara-=pes;
}

string getType()
{
    return this->getTipusBrossa();
}
*/
